### PR TITLE
Add onAction to bonify onBefore

### LINKED ISSUE #219: onBefore hook not runs onpopstate and on load()
onBefore hook not runs on "onpopstate" event (when user use "Back" button) and when load() method called.
I think it should be in docs, or if it is bug, it should be fixed.

### PR DESCRIPTION
**Pull request by mcaskill**
Resolves #219

The only missing feature for is a way to store the `$trigger` within the pushed state so as to be able to use it as the `$currentTarget` in the `onBefore()` function when `onPopState()` is called.

Currently with this change, any custom `onBefore()` declarations need to test for `$currentTarget` or suffer an error.

If this PR is accepted, please update the demos in your 'master' branch so as to reduce the differences.

### CODE REVIEW

**Comment by mcaskill** [comment]
After reading anew the issue referenced above, the code I've submitted, and how our project needs `onBefore` to be triggered and I think this PR is unnecessary. It shouldn't be merged and we should just update the documentation of the method to highlight the behaviour and suggest an alternative to the integrator.

Here's why:

In the majority of cases—like ours, @shrpne, and @nsciacca—we use the `onBefore` function to prepare the groundwork for _how_ the page animates based on _what_ we interacted with (the origin of the trigger, `$currentTarget`).

We click a link, `onBefore` is triggered, and we prepare stuff. SmoothState replaces the current page with a new one; which means the original DOM element we clicked is destroyed in favour of the new content. It makes sense that when the `onPopState` event is triggered, we can't naturally use the `onBefore` function.

If someone really needs for the `onBefore` function to be triggered during a `onPopState` event, we could think of an alternative that could be added to the documentation.

For example,
1. Move animation preparations from `onBefore` to `onStart`;
2. In the `onStart`, wrap the different animation "sets" in a `switch`/`case`;
3. From `onBefore`, resolve what link was clicked and which animation "set" it should apply and assign its unique identifier to a variable that is accessible from `onStart`;
4. Replace `$.smoothStateUtility.storePageIn` to also store the unique animation identifier;
5. Replace `$.smoothStateUtility.onPopState` to look up the cache and extract the animation identifier, then trigger the `onBefore` handler before the `load()`.

**Comment by nsciacca** [comment]
I see the onBefore as a way to unload/prepare the page to accept new content, not just for animating content differently.

For example, in my case I have several plugins that alter body classes & window event handlers based on the content loaded. When the content changes, I need to be able to unbind these event handles and alter the classes - regardless of whether a link was clicked or the browser navigation back/forward was used. Otherwise the page content goes into the smoothstate cache as if these plugins are loaded properly, but when the page is retrieved they don't function properly. The best way I've found is to destroy any behaviors before the page is stored.

Also, just for note, my links are in a header area and I'm manually attaching the smoothstate click behavior to them. I'm actually swapping only a selected portion of the page so my original DOM elements always persist.

My forked version is behind the master commit now by a few commits (going to update it, but it would be great if you kept yours as the PR). Let me know if you want to discuss more.

**Comment by mcaskill** [comment]
Good point. The project I'm working on also has a header area (outside of SmoothState's container) and we are having to deal with unbinding events when changing pages (I've arrived after much of the integration has been completed). Currently, our unbinding occurs in the `onStart()` handler but only for one particular page (the other pages haven't had unbinding implemented yet). Also, caching hasn't been activated so I am not aware of issues between `fetch()` and `onStart()`.

So basically, the `onBefore()` handler is used by `clickAnchor()` and `submitForm()`; maybe that handler should be renamed `onAction($currentTarget, $container)`; then add an `onBefore(request, $container)` early on in the `load()` function, before it calls `fetch()`.

This would probably help with the confusion of `onBefore`'s naming and would trigger with `onPopState`.

**Comment by mcaskill** [comment]
Closing this pull request because it has been inactive for over 600 days.
You may reopen this request if it is still useful.
